File: static/casting/sources/byte_model_tree_forest.c
/* byte_model_tree_forest.c — bagging: a forest of tiny decision trees.
 *
 * Training data: a 4-input boolean target (16 rows, true labels).
 * Corruption:    flip K random labels in training to simulate noise.
 * Single tree:   ID3 on the corrupted set; tests against the clean rows.
 * Forest:        F trees, each trained on a bootstrap sample (16 draws
 *                with replacement from the corrupted set). Predict by
 *                majority vote.
 *
 * The thesis: a single DT happily overfits the noise, so its test
 * accuracy drops below the noiseless ceiling. Trees trained on
 * different bootstrap samples overfit DIFFERENTLY, so majority voting
 * washes out idiosyncratic errors. Forest accuracy > single-tree
 * accuracy for any noisy dataset large enough to support diverse trees.
 *
 * Compile: cc -O2 -o byte_model_tree_forest byte_model_tree_forest.c -lm
 */

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#define N_FEATURES   4
#define N_SAMPLES    16
#define FOREST_SIZE  21
#define NOISE_FLIPS   3
#define MAX_NODES    64

typedef struct { int features[N_FEATURES]; int label; } Sample;
typedef struct { int is_leaf; int v; int f; int zc; int oc; } DTNode;

/* Per-tree node pool to avoid collisions between forest members. */
static DTNode g_pool[FOREST_SIZE + 1][MAX_NODES];
static int    g_count[FOREST_SIZE + 1];

static double entropy(const Sample *s, const int *ix, int n) {
    int pos = 0;
    for (int i = 0; i < n; ++i) if (s[ix[i]].label > 0) pos++;
    int neg = n - pos;
    if (pos == 0 || neg == 0) return 0.0;
    double p = (double)pos/n, q = (double)neg/n;
    return -p*log(p) - q*log(q);
}
static int best_feature(const Sample *s, const int *ix, int n,
                        const int *avail) {
    double base = entropy(s, ix, n);
    int best = -1; double best_gain = -1.0;
    for (int f = 0; f < N_FEATURES; ++f) {
        if (!avail[f]) continue;
        int z[N_SAMPLES], o[N_SAMPLES]; int nz = 0, no = 0;
        for (int i = 0; i < n; ++i) {
            if (s[ix[i]].features[f]) o[no++] = ix[i];
            else                      z[nz++] = ix[i];
        }
        double after = 0.0;
        if (nz) after += (double)nz/n * entropy(s, z, nz);
        if (no) after += (double)no/n * entropy(s, o, no);
        double gain = base - after;
        if (gain > best_gain) { best_gain = gain; best = f; }
    }
    return best;
}

/* Random-subspace variant: at each split, only let `max_features` of the
 * currently-available features compete. Key trick for diverse forests. */
static int best_feature_rand(const Sample *s, const int *ix, int n,
                             const int *avail, int max_features) {
    int cand[N_FEATURES]; int nc = 0;
    for (int f = 0; f < N_FEATURES; ++f) if (avail[f]) cand[nc++] = f;
    if (nc == 0) return -1;
    for (int i = nc - 1; i > 0; --i) {
        int j = rand() % (i + 1);
        int tmp = cand[i]; cand[i] = cand[j]; cand[j] = tmp;
    }
    int keep = nc < max_features ? nc : max_features;
    int subset[N_FEATURES] = {0};
    for (int i = 0; i < keep; ++i) subset[cand[i]] = 1;
    return best_feature(s, ix, n, subset);
}
static int emit_leaf(int tree, int v) {
    int idx = g_count[tree]++;
    g_pool[tree][idx] = (DTNode){1, v, -1, -1, -1};
    return idx;
}
static int majority(const Sample *s, const int *ix, int n) {
    int pos = 0;
    for (int i = 0; i < n; ++i) if (s[ix[i]].label > 0) pos++;
    return (pos >= n - pos) ? +1 : -1;
}
static int build(int tree, const Sample *s, const int *ix, int n,
                 const int *avail, int max_features) {
    int pos = 0;
    for (int i = 0; i < n; ++i) if (s[ix[i]].label > 0) pos++;
    if (pos == 0) return emit_leaf(tree, -1);
    if (pos == n) return emit_leaf(tree, +1);
    int any = 0;
    for (int f = 0; f < N_FEATURES; ++f) if (avail[f]) { any = 1; break; }
    if (!any) return emit_leaf(tree, majority(s, ix, n));
    int f = (max_features >= N_FEATURES)
        ? best_feature(s, ix, n, avail)
        : best_feature_rand(s, ix, n, avail, max_features);
    if (f < 0) return emit_leaf(tree, majority(s, ix, n));
    int z[N_SAMPLES], o[N_SAMPLES]; int nz = 0, no = 0;
    for (int i = 0; i < n; ++i) {
        if (s[ix[i]].features[f]) o[no++] = ix[i];
        else                      z[nz++] = ix[i];
    }
    int na[N_FEATURES]; memcpy(na, avail, sizeof(na)); na[f] = 0;
    int slot = g_count[tree]++;
    int zc = build(tree, s, z, nz, na, max_features);
    int oc = build(tree, s, o, no, na, max_features);
    g_pool[tree][slot] = (DTNode){0, 0, f, zc, oc};
    return slot;
}
static int eval_tree(int tree, int root, const int *feats) {
    int idx = root;
    while (!g_pool[tree][idx].is_leaf)
        idx = feats[g_pool[tree][idx].f] ? g_pool[tree][idx].oc : g_pool[tree][idx].zc;
    return g_pool[tree][idx].v;
}

/* --- data helpers ----------------------------------------------------- */

static void generate_clean(Sample *s, uint64_t table) {
    for (int row = 0; row < N_SAMPLES; ++row) {
        s[row].label = ((table >> row) & 1) ? +1 : -1;
        for (int f = 0; f < N_FEATURES; ++f)
            s[row].features[f] = (row >> (N_FEATURES - 1 - f)) & 1;
    }
}
static void flip_labels(Sample *s, int k) {
    int used[N_SAMPLES] = {0};
    for (int i = 0; i < k; ++i) {
        int r;
        do { r = rand() % N_SAMPLES; } while (used[r]);
        used[r] = 1;
        s[r].label = -s[r].label;
    }
}

/* --- runs ------------------------------------------------------------- */

static void run_case(const char *name, uint64_t table) {
    Sample clean[N_SAMPLES], train[N_SAMPLES];
    generate_clean(clean, table);
    memcpy(train, clean, sizeof train);
    flip_labels(train, NOISE_FLIPS);

    /* Which rows are corrupted? */
    int flipped_mask = 0;
    for (int i = 0; i < N_SAMPLES; ++i)
        if (train[i].label != clean[i].label) flipped_mask |= (1 << i);

    /* single tree: full feature set, no bootstrap */
    g_count[0] = 0;
    int ix[N_SAMPLES]; for (int i = 0; i < N_SAMPLES; ++i) ix[i] = i;
    int avail[N_FEATURES] = {1,1,1,1};
    int root_single = build(0, train, ix, N_SAMPLES, avail, N_FEATURES);
    int single_correct = 0;
    for (int i = 0; i < N_SAMPLES; ++i)
        if (eval_tree(0, root_single, clean[i].features) == clean[i].label)
            single_correct++;

    /* forest: F trees, each with bootstrap + random 2-of-4 feature subspace */
    int roots[FOREST_SIZE];
    for (int t = 0; t < FOREST_SIZE; ++t) {
        g_count[t + 1] = 0;
        int boot_ix[N_SAMPLES];
        for (int i = 0; i < N_SAMPLES; ++i) boot_ix[i] = rand() % N_SAMPLES;
        int aa[N_FEATURES] = {1,1,1,1};
        roots[t] = build(t + 1, train, boot_ix, N_SAMPLES, aa, 2);
    }
    int forest_correct = 0;
    for (int i = 0; i < N_SAMPLES; ++i) {
        int vote = 0;
        for (int t = 0; t < FOREST_SIZE; ++t)
            vote += eval_tree(t + 1, roots[t], clean[i].features);
        int y = vote >= 0 ? +1 : -1;
        if (y == clean[i].label) forest_correct++;
    }

    printf("target %-16s  flipped rows: 0x%04x (%d / %d)\n",
           name, flipped_mask, NOISE_FLIPS, N_SAMPLES);
    printf("  single tree          test acc: %2d / %d\n",
           single_correct, N_SAMPLES);
    printf("  forest  (F=%d vote)  test acc: %2d / %d\n\n",
           FOREST_SIZE, forest_correct, N_SAMPLES);
}

int main(void) {
    srand(42);
    printf("bagging: single tree vs %d-tree forest on noisy 4-input targets\n"
           "(training labels corrupted by %d random flips; test on clean truth)\n\n",
           FOREST_SIZE, NOISE_FLIPS);
    run_case("4-MAJ (>=3)",   0xe880);
    run_case("4-threshold 2", 0xfee8);
    run_case("4-OR",          0xfffe);
    run_case("4-AND",         0x8000);
    printf("Typical result: forest accuracy >= single-tree accuracy. Each\n"
           "bootstrap tree overfits a DIFFERENT subset of noise; majority\n"
           "vote washes out idiosyncratic errors. On low-noise targets the\n"
           "gap narrows — there is nothing for bagging to smooth over.\n");
    return 0;
}

File: static/casting/js/byte_model_tree_forest.js
/* Casting: byte_model_tree_forest — bagging with random subspaces.
 * Mirrors static/casting/sources/byte_model_tree_forest.c.
 */
(function () {
  const N_FEATURES = 4;
  const N_SAMPLES  = 16;
  const FOREST_SIZE = 21;
  const NOISE_FLIPS = 3;

  /* Tiny LCG so we get reproducible "random" behaviour per-run. */
  let rngState = 42;
  function rnd() {
    rngState = (rngState * 1103515245 + 12345) & 0x7fffffff;
    return rngState;
  }
  function rndRange(n) { return rnd() % n; }
  function rndReset(seed) { rngState = seed; }

  function entropy(samples, ix) {
    let pos = 0;
    for (const i of ix) if (samples[i].label > 0) pos++;
    const neg = ix.length - pos;
    if (pos === 0 || neg === 0) return 0;
    const p = pos / ix.length, q = neg / ix.length;
    return -p * Math.log(p) - q * Math.log(q);
  }
  function bestFeature(samples, ix, avail) {
    const base = entropy(samples, ix);
    let best = -1, bestGain = -1;
    for (let f = 0; f < N_FEATURES; ++f) {
      if (!avail[f]) continue;
      const z = [], o = [];
      for (const i of ix) (samples[i].features[f] ? o : z).push(i);
      let after = 0;
      if (z.length) after += z.length / ix.length * entropy(samples, z);
      if (o.length) after += o.length / ix.length * entropy(samples, o);
      const gain = base - after;
      if (gain > bestGain) { bestGain = gain; best = f; }
    }
    return best;
  }
  function bestFeatureRand(samples, ix, avail, maxFeat) {
    const cand = [];
    for (let f = 0; f < N_FEATURES; ++f) if (avail[f]) cand.push(f);
    if (!cand.length) return -1;
    for (let i = cand.length - 1; i > 0; --i) {
      const j = rndRange(i + 1);
      [cand[i], cand[j]] = [cand[j], cand[i]];
    }
    const keep = Math.min(cand.length, maxFeat);
    const subset = new Array(N_FEATURES).fill(0);
    for (let i = 0; i < keep; ++i) subset[cand[i]] = 1;
    return bestFeature(samples, ix, subset);
  }

  function majority(samples, ix) {
    let pos = 0;
    for (const i of ix) if (samples[i].label > 0) pos++;
    return pos >= ix.length - pos ? +1 : -1;
  }

  function build(nodes, samples, ix, avail, maxFeat) {
    let pos = 0;
    for (const i of ix) if (samples[i].label > 0) pos++;
    if (pos === 0)          { nodes.push({ leaf: true, v: -1 }); return nodes.length - 1; }
    if (pos === ix.length)  { nodes.push({ leaf: true, v: +1 }); return nodes.length - 1; }
    if (!avail.some(v => v)){ nodes.push({ leaf: true, v: majority(samples, ix) }); return nodes.length - 1; }
    const f = (maxFeat >= N_FEATURES)
      ? bestFeature(samples, ix, avail)
      : bestFeatureRand(samples, ix, avail, maxFeat);
    if (f < 0) { nodes.push({ leaf: true, v: majority(samples, ix) }); return nodes.length - 1; }
    const z = [], o = [];
    for (const i of ix) (samples[i].features[f] ? o : z).push(i);
    const na = avail.slice(); na[f] = 0;
    const slot = nodes.length;
    nodes.push(null);
    const zc = build(nodes, samples, z, na, maxFeat);
    const oc = build(nodes, samples, o, na, maxFeat);
    nodes[slot] = { leaf: false, f, zc, oc };
    return slot;
  }
  function evalTree(nodes, root, feats) {
    let idx = root;
    while (!nodes[idx].leaf)
      idx = feats[nodes[idx].f] ? nodes[idx].oc : nodes[idx].zc;
    return nodes[idx].v;
  }

  function generateClean(table) {
    const s = [];
    for (let row = 0; row < N_SAMPLES; ++row) {
      const features = [];
      for (let f = 0; f < N_FEATURES; ++f)
        features.push((row >> (N_FEATURES - 1 - f)) & 1);
      s.push({ features, label: ((table >> row) & 1) ? +1 : -1 });
    }
    return s;
  }
  function flipLabels(train, k) {
    const used = new Set();
    while (used.size < k) {
      const r = rndRange(N_SAMPLES);
      if (used.has(r)) continue;
      used.add(r);
      train[r] = Object.assign({}, train[r], { label: -train[r].label });
    }
    return used;
  }

  function runCase(lines, name, table) {
    const clean = generateClean(table);
    const train = clean.map(s => Object.assign({}, s));
    const flipped = flipLabels(train, NOISE_FLIPS);
    let flipMask = 0;
    for (const i of flipped) flipMask |= (1 << i);

    /* single tree */
    const singleNodes = [];
    const ix = clean.map((_, i) => i);
    const rootSingle = build(singleNodes, train, ix, [1,1,1,1], N_FEATURES);
    let singleCorrect = 0;
    for (const s of clean)
      if (evalTree(singleNodes, rootSingle, s.features) === s.label) singleCorrect++;

    /* forest */
    const trees = [];
    for (let t = 0; t < FOREST_SIZE; ++t) {
      const bootIx = [];
      for (let i = 0; i < N_SAMPLES; ++i) bootIx.push(rndRange(N_SAMPLES));
      const nodes = [];
      const root = build(nodes, train, bootIx, [1,1,1,1], 2);
      trees.push({ nodes, root });
    }
    let forestCorrect = 0;
    for (const s of clean) {
      let vote = 0;
      for (const t of trees) vote += evalTree(t.nodes, t.root, s.features);
      const y = vote >= 0 ? +1 : -1;
      if (y === s.label) forestCorrect++;
    }

    const mask = flipMask.toString(16).padStart(4, '0');
    lines.push(`target ${name.padEnd(16)}  flipped rows: 0x${mask} (${NOISE_FLIPS} / ${N_SAMPLES})`);
    lines.push(`  single tree          test acc: ${String(singleCorrect).padStart(2)} / ${N_SAMPLES}`);
    lines.push(`  forest  (F=${FOREST_SIZE} vote)  test acc: ${String(forestCorrect).padStart(2)} / ${N_SAMPLES}`);
    lines.push('');
  }

  const CASES = [
    { name: '4-MAJ (>=3)',   table: 0xe880 },
    { name: '4-threshold 2', table: 0xfee8 },
    { name: '4-OR',          table: 0xfffe },
    { name: '4-AND',         table: 0x8000 },
  ];

  function run() {
    rndReset(42);
    const lines = [];
    lines.push(`bagging: single tree vs ${FOREST_SIZE}-tree forest on noisy 4-input targets`);
    lines.push(`(training labels corrupted by ${NOISE_FLIPS} random flips; test on clean truth)`);
    lines.push('');
    for (const c of CASES) runCase(lines, c.name, c.table);
    lines.push('Typical result: forest accuracy >= single-tree accuracy. Each');
    lines.push('bootstrap + random-feature-subset tree overfits a DIFFERENT subset');
    lines.push('of noise; majority vote washes out idiosyncratic errors. The effect');
    lines.push('is modest on tiny noise-free targets — not much to smooth over.');
    return lines.join('\n');
  }

  window.Casting_byte_model_tree_forest = { run };
})();
